Support Signer Interface for signing SD-JWTs

cryptphil opened the pull request:
This PR adds support for external signer implementations used to sign SD-JWTs by introducing an SD-JWT signer interface.
Therefore, the `createCredential` method can now also be called by providing a signer instance instead of a private key.

Closes #5.

Commit:
Support custom signer implementations for signing SD-JWTs

Signed-off-by: Philipp-Florens Lehwalder <philipp.lehwalder@neosfer.com>

Commit:
Add test for StaticJWSSigner

Signed-off-by: Philipp-Florens Lehwalder <philipp.lehwalder@neosfer.com>

markuskreusch commented:
Thanks for the contribution. Looks promising. A thing we had on our end that would be helpful is, that the signer interface could generate the base header by itself, instead of just returning algorithm and kid. This would allow for more flexibility when generating the SD-JWT. Some signature mechanisms might want to use kid but others will not do that. We for example need a way to set the x5c header and and omit the kid.

So I am looking for something like:

```kotlin
interface SdJwtSigner {
    fun baseHeader(): JWSHeader.Builder
    fun jwsSigner(): JWSSigner
}
```

We could discuss which role the SdJwtHeader plays in this. I would refrain from duplicating all the possible header entries there that might be set and setting it this way. Setting x5c, kid, sigT (this is a custom JAdES header) and others belongs to the signing implementation IMHO. SdJwtHeader only serves the purpose of setting SD-JWT specific header fields.

An additional remark is, that I am not that happy with the fact that it extends JWSSigner instead of returning one. Basically the thing only signs SD-JWTs created with the library but could be used in other places where a JWSSigner is required when it extends that. The additional header values would then be ignored though. Thus returning a JWSSigner would be better IMHO to avoid confusion.

When it no longer is a JWSSigner it should not be called *JWSSigner any more. SdJwtSigner would work, as well as SigningConfiguration or SignerProvider or SignatureProvider. Just some thoughts, maybe you can come up with a better name.

 I dislike, that we now have two methods buildJWT one for a key and one for a SdJwtSigner. I would do it the following way:
- Provide one method createCredential that accepts a signer, this calls all other methods necessary to create the SD-JWT
- Create in implementation of SdJwtSigner "KeyBasedSdJwtSigner", that works with a key to create a signature.The KeyBasedSdJwtSigner can be public so it can be constructed directly. This may help when external callers want to invoke the methods sometimes with a key based approach and other times with another signing implementation.
- Provide one method createCredential that accepts a key for backwards compatibility. It only invokes the other createCredential method accepting a signer with an instance of KeyBasedSdJwtSigner. (@fabian-hk Please check if this is needed, if we can drop this and require that lib users call the signer based method directly even better).

cryptphil commented:
Thanks for your quick and constructive feedback, @markuskreusch! I will address your points as soon as possible, in the meantime I have set this PR as a draft.

Commit:
Refactor StaticJWSSigner to SdJwtSigner

SdJwtSigner returns a JWSHeader and JWSSigner
Provide KeyBasedSdJwtSigner implementation

Signed-off-by: Philipp-Florens Lehwalder <philipp.lehwalder@neosfer.com>

cryptphil commented:
I addressed your feedback and introduced a `SdJwtSigner` interface @markuskreusch.
The `createCredential` method, which expects a private key, also still exists and calls the new one with a `KeyBasedSdJwtSigner` but I added a comment that this is actually a legacy method now.

markuskreusch commented:
@cryptphil Exactly what I had in mind.

One small thing: The KeyBasedSdJwtSigner now always returns the same header builder instance. This will work if a new instance is created per credential creation, but will not work if the same instance is reused. The baseHeader method should return a new builder each time it is invoked.

You could:
- Change the type of the header field from JWSHeader.Builder to JWSHeader
- Create a pre filled JWSHeader in the init block
- Call JWSHeader.Builder(header) in the base header method

Apart from that LGTM.

Commit:
Fix KeyBasedSdJwtSigner returning same JWSHeader Builder instance

Signed-off-by: Philipp-Florens Lehwalder <philipp.lehwalder@neosfer.com>

cryptphil commented:
@markuskreusch Thanks, good catch! I fixed it using your second proposed suggestion such that the return type is still JWSHeader.Builder.

Commit:
Merge branch 'main' into support-signer-interface

Signed-off-by: Philipp-Florens Lehwalder <philipp.lehwalder@neosfer.com>

# Conflicts:
#	src/main/kotlin/org/sd_jwt/SdJwt.kt
#	src/test/kotlin/org/sd_jwt/Utils.kt

Commit:
Remove unnecessary function inlining

Signed-off-by: Philipp-Florens Lehwalder <philipp.lehwalder@neosfer.com>

cryptphil commented:
I have now merged the changes from PR #4 into mine and this one would be ready to review again @fabian-hk

fabian-hk commented:
From my point of view we can drop the old method without the signer interface. As far as I know you two are the only users of the library right now. I will look into the PR as soon as possible.

fabian-hk commented on src/main/kotlin/org/sd_jwt/SdJwt.kt:
```diff
@@ -155,8 +148,13 @@ fun createSdClaims(
 /**
- * This method creates a SD-JWT credential that contains the claims
- * passed to the method and is signed with the issuer's key.
+ * [LEGACY METHOD]
```
The createCredential method that processes data classes should not be a legacy method. I do not plan to remove this method. I would like you to change the issuerKey: JWK parameter to signer: SdJwtSigner. I know this is a breaking change but there was no official release yet so I think this is okay. Other than that the PR looks good and is ready to merge I think.

cryptphil replied:
Alright thanks, I have removed the "old" method such that `createCredential` always expects a signer.

fabian-hk replied:
I think there was a misunderstanding. I do want to keep two createCredential methods. One is to create a credential from a JSONObject using a signer and the other one creates a credential from a data class also using a signer. I only wanted you to change the legacy method to use a signer and remove the legacy comment.

cryptphil replied:
Ah okay, I understand; should be fixed now!

cryptphil replied:
Ah okay, I understand; should be fixed now! Please have another look.

Commit:
Refactor createCredential expecting claim data class to use signer instance

Signed-off-by: Philipp-Florens Lehwalder <philipp.lehwalder@neosfer.com>

fabian-hk commented on src/main/kotlin/org/sd_jwt/SdJwt.kt:
```diff
@@ -166,13 +161,12 @@ fun createSdClaims(
+    signer: SdJwtSigner,
     holderPubKey: JWK? = null,
     discloseStructure: T? = null,
     sdJwtHeader: SdJwtHeader = SdJwtHeader(),
```
I guess we don't need the sdJwtHeader anymore because the header can be set via the signer, right? If so, can you remove everything related to that please.

cryptphil replied:
Yes, that is possible, but it also required a change at the `SdJwtSigner`; it now can be provided with a `SdJwtHeader` and will then return the complete `JWSHeader` including the potential typ and cty header parameters using the `sdJwtHeader()` method. Let me know what you think of this.

fabian-hk replied:
I think it's better now but would it be possible to just pass a ``JWSHeader`` object to the signer? That would give the user complete flexibility over the header. Or do you think that does not make sense?

fabian-hk replied:
The ``JWSHeader`` object should be an optional argument.

cryptphil replied:
I don't think there is really a need for specifying more parameters than the SD-JWT custom ones for the type and cty for the `KeyBasedSdJwtSigner`. If one needs full control over the header, one can still provide its own implementation of the `SdJwtSigner`, the key-based implementation is anyway more or less just a simple "hack" to still use a JWK.

Additionally, when providing a `JWSHeader` or builder, it wouldn't readily allow the `KeyBasedSdJwtSigner` to set the algorithm in the header by itself (as it already needs to be specified when providing a header).

Commit:
Create SD-JWT header in SdJwtSigner

Signed-off-by: Philipp-Florens Lehwalder <philipp.lehwalder@neosfer.com>

fabian-hk commented on src/main/kotlin/org/sd_jwt/SdJwt.kt:
```diff
@@ -408,7 +387,10 @@ inline fun internalCreatePresentation(
             .build()
-        presentation += SEPARATOR + buildJWT(holderBindingJwtPayload, holderKey)
+
+        val keyBasedSdJwtSigner = holderKey?.let { KeyBasedSdJwtSigner(it) }
```
Should we not also use a signer for the holder key? Or is this not a practical use case? But I think it would be important for holder keys stored in HSMs.

cryptphil replied:
I agree, having a signer interface for the holder JWK would also be useful for implementors using the sd-jwt library as the holder.

I changed it now by adding a method to the SdJwtSigner for getting the public JWK.

Commit:
Use signer interface instead of holder JWK for presentations

Signed-off-by: Philipp-Florens Lehwalder <philipp.lehwalder@neosfer.com>

fabian-hk commented:
The PR looks good now. Thank you very much for your contribution!
